### PR TITLE
Store long term credentials in system keychain

### PR DESCRIPTION
**Pull request by ecneladis**
- [x] update `README.md` with described keychain feature

### CODE REVIEW

File: awsmfa/config.py [hunk 1]
```diff
@@ -25,8 +26,5 @@ def initial_setup(logger, config, config_path):
     if aws_secret_access_key is None or aws_secret_access_key == "":
         log_error_and_exit(logger, "You must supply aws_secret_access_key")
 
-    config.add_section(profile_name)
-    config.set(profile_name, 'aws_access_key_id', aws_access_key_id)
-    config.set(profile_name, 'aws_secret_access_key', aws_secret_access_key)
-    with open(config_path, 'w') as configfile:
-        config.write(configfile)
+    keyring.set_password("aws:access_key_id", profile_name, aws_access_key_id)
+    keyring.set_password("aws:secret_access_key", profile_name, awsgetpass.getpass_secret_access_key)
```
Comment (pmoust) [inline awsmfa/config.py]: just `aws_secret_access_key` here, you've already gotten the value with `getpass` earlier

File: setup.py [hunk 2]
```diff
@@ -24,5 +24,8 @@
         ],
     },
     url='https://github.com/broamski/aws-mfa',
-    install_requires=['boto3']
+    install_requires=[
```
Comment (pmoust) [inline setup.py:27]: aren't `'dbus-python', 'secretstorage'` required as well to support D-Bus?

**Review by pmoust — CHANGES_REQUESTED** [review]

File: awsmfa/config.py (hunk 1 above)
Comment (ecneladis) [reply awsmfa/config.py]: This code is responsible for putting credentials into keychain, not for getting them.

Comment (pmoust) [reply awsmfa/config.py]: Yeah but check the value you are trying to set. it should be `aws_secret_access_key` as mentioned above (and not `awsgetpass.getpass_secret_access_key`)

Comment (ecneladis) [reply awsmfa/config.py]: Got it! accidental string paste.

**Comment by gservat** [comment]
Will aws-mfa continue to work for those not on OS/X?

**Comment by gservat** [comment]
Actually @pmoust mentioned via email that it's cross platform.

**Comment by ecneladis** [comment]
@gservat `keyring` lib supports MacOS, Linux and Windows - https://github.com/jaraco/keyring#what-is-python-keyring-lib

File: setup.py (hunk 2 above)
Comment (ecneladis) [reply setup.py:27]: Resolved in f558bc5bb5c367e6071baada46a5567db24c6d07

**Review by pmoust — CHANGES_REQUESTED** [review]
Could you add a command line argument (like `--no-keychain`) to set/get long-term creds from disk instead?

This is to allow users the (unsafe, non default) method of having long-term keys on disk as before.

Some Linux users may not be using D-Bus (i.e. folks working on headless VMs etc)

**Review by pmoust — APPROVED** [review]
LGTM 

Before merging, 'd like someone with Gnome (or other d-bus enabled thingie) to test it as well please ./cc @elastic/cloud-sre

**Comment by ecneladis** [comment]
- [ ] update `README.md` with described keychain feature

**Comment by mattfield** [comment]
Can we get a series of QA steps for this please?

**Comment by mattfield** [comment]
Confirmed that this works with a compatible keyring like `gnome-keyring`

**Comment by ecneladis** [comment]
`README.md` updated

File: README.md [hunk 3]
```diff
@@ -167,6 +169,10 @@ Usage
 --role-session-name ROLE_SESSION_NAME
                         Friendly session name required when using --assume-
                         role. By default, this is your local username.
+--token TOKEN, --mfa-token TOKEN
+                        Provide MFA token as an argument
+--no-keychain           Do not use system keychain to store long term
```
Comment (pmoust) [inline README.md]: s/to store/to store or retrieve/

File: awsmfa/__init__.py [hunk 4]
```diff
@@ -83,6 +84,10 @@ def main():
                         type=str,
                         help="Provide MFA token as an argument",
                         required=False)
+    parser.add_argument('--no-keychain',
+                        action="store_true",
+                        help="Do not use system keychain to store long term credentials",
```
Comment (pmoust) [inline awsmfa/__init__.py]: as above

File: README.md [hunk 5]
```diff
@@ -29,9 +29,11 @@ Option 2
 Credentials File Setup
 ----------------------
 
-In a typical AWS credentials file (located at `~/.aws/credentials`), credentials are stored in sections, denoted by a pair of brackets: `[]`. The `[default]` section stores your default credentials. You can store multiple sets of credentials using different profile names. If no profile is specified, the `[default]` section is always used.
+By default long term credentials are stored in system keychain (using [keyring library](https://pypi.org/project/keyring/)), you can store and use credentials from credentials file (located at `~/.aws/credentials`) by running with `--no-keychain` command line argument. Only short credentials are stored AWS credentials file.
```
Comment (pmoust) [inline README.md]: > By default long term credentials are stored in system keychain (using [keyring library](https://pypi.org/project/keyring/)), only short term credentials are stored in `~/.aws/credentials`.
> It is possible to _not_ use system keychain to store and retrieve long term credentials  by running **aws-mfa** with the `--no-keychain` command line flag. When using the `--no-keychain` flag, long term credentials are stored in and retrieved from `~/.aws/credentials` (*NOT RECOMMENDED*).

**Review by pmoust — CHANGES_REQUESTED** [review]

**Review by pmoust — APPROVED** [review]

**Comment by matschaffer** [comment]
@pmoust @ecneladis maybe I'm just missing it, but how does the id/secret get _into_ the OS keychain in the first place? Doesn't seem clear from the docs.

**Comment by pmoust** [comment]
You are right, we need an explicit reference to `aws-mfa --setup`

**Pull request merged**